// egg-logger/lib/egg/loggers.js
'use strict';

const path = require('path');
const assert = require('assert');
const debug = require('debug')('egg:logger');
const utils = require('../utils');
const Logger = require('./logger');
const ErrorLogger = require('./error_logger');
const CustomLogger = require('./custom_logger');
// const cluster = require('cluster');

const defaults = {
  env: 'default',
  type: '',
  dir: '',
  encoding: 'utf8',
  level: 'INFO',
  consoleLevel: 'NONE',
  outputJSON: false,
  buffer: true,
  appLogName: '',
  coreLogName: '',
  agentLogName: '',
  errorLogName: '',
};

/**
 * Logger Manager, accord config to create multi loggers.
 */

class Loggers extends Map {

  /**
   * @constructor
   * @param  {Object} config - egg app config
   * - logger
   *   - {String} env - egg app runtime env string, detail please see `app.config.env`
   *   - {String} type - current process type, `application` or `agent`
   *   - {String} dir - log file dir
   *   - {String} [encoding = utf8] - log string encoding
   *   - {String} [level = INFO] - file log level
   *   - {String} [consoleLevel = NONE] - console log level
   *   - {Boolean} [outputJSON = false] - send JSON log or not
   *   - {Boolean} [buffer = true] - use {@link FileBufferTransport} or not
   *   - {String} appLogName - egg app file logger name
   *   - {String} coreLogName - egg core file logger name
   *   - {String} agentLogName - egg agent file logger name
   *   - {String} errorLogName - err common error logger name
   *   - {String} eol - end of line char
   * - customLogger
   */
  // app.config参数
  constructor(config) {
    super();
    // console.log('egg-logger', process.pid, cluster.isWorker, cluster.isMaster);
    // egg内有一些默认logger的设置
    const loggerConfig = utils.assign({}, defaults, config.logger);
    const customLoggerConfig = config.customLogger;

    debug('Init loggers with options %j', loggerConfig);
    assert(loggerConfig.type, 'should pass config.logger.type');
    assert(loggerConfig.dir, 'should pass config.logger.dir');
    assert(loggerConfig.appLogName, 'should pass config.logger.appLogName');
    assert(loggerConfig.coreLogName, 'should pass config.logger.coreLogName');
    assert(loggerConfig.agentLogName, 'should pass config.logger.agentLogName');
    assert(loggerConfig.errorLogName, 'should pass config.logger.errorLogName');

    const errorLogger = new ErrorLogger(utils.assign({}, loggerConfig, {
      // egg里config.default.js设置的dir和errorLogName
      file: path.join(loggerConfig.dir, loggerConfig.errorLogName),
    }));
    // logger本身是个map。其errorLogger也是个map。打印的时候基于map.values(),会轮询每个实例
    this.set('errorLogger', errorLogger);
    // config.logger.type。egg的logger文件中，针对agent和worker进程做了type标记。
    if (loggerConfig.type === 'agent') {
      // agent进程一个。
      const logger = new Logger(utils.assign({}, loggerConfig, {
        file: path.join(loggerConfig.dir, loggerConfig.agentLogName),
      }));
      this.set('logger', logger);

      const coreLogger = new Logger(utils.assign({}, loggerConfig, loggerConfig.coreLogger, {
        file: path.join(loggerConfig.dir, loggerConfig.agentLogName),
      }));
      this.set('coreLogger', coreLogger);
    } else {
      const logger = new Logger(utils.assign({}, loggerConfig, {
        file: path.join(loggerConfig.dir, loggerConfig.appLogName),
      }));
      this.set('logger', logger);

      const coreLogger = new Logger(utils.assign({}, loggerConfig, loggerConfig.coreLogger, {
        file: path.join(loggerConfig.dir, loggerConfig.coreLogName),
      }));
      this.set('coreLogger', coreLogger);
    }
    // 增加 'error', 'warn', 'info', 'debug'方法。
    for (const name in customLoggerConfig) {
      const logger = new CustomLogger(utils.assign({}, loggerConfig, customLoggerConfig[name]));
      this.set(name, logger);
    }
  }

  /**
   * Disable console logger
   */
  disableConsole() {
    for (const logger of this.values()) {
      logger.disable('console');
    }
  }

  reload() {
    for (const logger of this.values()) {
      logger.reload();
    }
  }

  /**
   * Add a logger
   * @param {String} name - logger name
   * @param {Logger} logger - Logger instance
   */
  set(name, logger) {
    if (this.has(name)) {
      return;
    }

    // redirect ERROR log to errorLogger, except errorLogger itself
    if (name !== 'errorLogger') {
      logger.redirect('error', this.errorLogger);
    }
    this[name] = logger;
    super.set(name, logger);
  }

}

module.exports = Loggers;
