// egg-schedule/lib/strategy/base.js
'use strict';

module.exports = class BaseStrategy {
  constructor(schedule, agent, key) {
    this.agent = agent;
    this.key = key;
    this.schedule = schedule;
    this.logger = this.agent.loggers.scheduleLogger;
  }

  start() {}

  sendOne(...args) {
    /* istanbul ignore next */
    if (this.agent.schedule.closed) {
      this.logger.warn(`${this.key} skip due to schedule closed`);
      return;
    }
    const id = this.getSeqId();
    this.logger.info(`[${id}] ${this.key} triggered, send random by agent`);
    // sendRandom如是worker进程，则发给agent进程。
    // sendRandom如是agent进程，则随机发给一个worker进程。
    this.agent.messenger.sendRandom('egg-schedule', { key: this.key, id, args });
  }

  sendAll(...args) {
    /* istanbul ignore next */
    if (this.agent.schedule.closed) {
      this.logger.warn(`${this.key} skip due to schedule closed`);
      return;
    }
    const id = this.getSeqId();
    this.logger.info(`[${id}] ${this.key} triggered, send all by agent`);
    // 发给所有worker进程。
    this.agent.messenger.send('egg-schedule', { key: this.key, id, args });
  }

  getSeqId() {
    return `${Date.now()}${process.hrtime().join('')}`;
  }
};
