// egg-schedule/lib/schedule.js
'use strict';

const STRATEGY = Symbol('strategy');
const STRATEGY_INSTANCE = Symbol('strategy_instance');
const loadSchedule = require('./load_schedule');

module.exports = class Schedule {
  constructor(agent) {
    this.agent = agent;
    this[STRATEGY] = new Map();
    this[STRATEGY_INSTANCE] = new Map();
    this.closed = false;
  }

  /**
   * register a custom Schedule Strategy
   * @param {String} type - strategy type
   * @param {Strategy} clz - Strategy class
   */
  use(type, clz) {
    // 类型，目前就2个。all/worker
    this[STRATEGY].set(type, clz);
  }

  init() {
    // 所有定时任务、包含插件，框架，应用的定时任务项。
    const scheduleItems = loadSchedule(this.agent);

    for (const k of Object.keys(scheduleItems)) {
      const { key, schedule } = scheduleItems[k];
      // 定时任务schedule配置的type。
      const type = schedule.type;
      if (schedule.disable) continue;
      // all/worker的strategy
      const Strategy = this[STRATEGY].get(type);
      if (!Strategy) {
        const err = new Error(`schedule type [${type}] is not defined`);
        err.name = 'EggScheduleError';
        throw err;
      }
      // Strategy类型实例，new出任务实例。
      const instance = new Strategy(schedule, this.agent, key);
      this[STRATEGY_INSTANCE].set(key, instance);
    }
  }

  start() {
    this.closed = false;
    // start所有任务实例。调起start方法。
    for (const instance of this[STRATEGY_INSTANCE].values()) {
      instance.start();
    }
  }

  close() {
    this.closed = true;
  }
};
